--- a/gdal/configure.ac
+++ b/gdal/configure.ac
@@ -1230,10 +1230,11 @@ else
 
   if test "x$with_proj" = "xyes" -o "x$with_proj" = "x"; then
     ORIG_LIBS="$LIBS"
-    LIBS="-lproj $ORIG_LIBS"
-    AC_LANG_PUSH([C++])
-    AC_CHECK_LIB(proj,proj_create_from_wkt,PROJ_FOUND=yes,PROJ_FOUND=no,)
-    AC_LANG_POP([C++])
+    PKG_PROG_PKG_CONFIG([0.21])
+    PKG_CHECK_MODULES([PROJ], [proj >= 6.0.0], [PROJ_FOUND=yes], [PROJ_FOUND=no])
+    LIBS="$PROJ_LIBS $ORIG_LIBS"
+    CFLAGS="$PROJ_CFLAGS $CFLAGS"
+    PROJ_INCLUDE="$PROJ_CFLAGS"
     if test "$PROJ_FOUND" = "no"; then
         AC_LANG_PUSH([C++])
         AC_CHECK_LIB(proj,internal_proj_create_from_wkt,PROJ_FOUND=yes,PROJ_FOUND=no,)
@@ -1839,10 +1840,10 @@ elif test "$with_png" = "internal" ; then
   echo "using internal png code."
 
 else
-
+  PKG_CHECK_MODULES([PNG], [libpng >= 1.0.0],,)
   PNG_SETTING=external
-  LIBS="-L$with_png -L$with_png/lib -lpng $LIBS"
-  EXTRA_INCLUDES="-I$with_png -I$with_png/include $EXTRA_INCLUDES"
+  LIBS="$PNG_LIBS $LIBS"
+  EXTRA_INCLUDES="$PNG_CFLAGS $EXTRA_INCLUDES"
 
   echo "using libpng from $with_png."
 
@@ -1978,21 +1979,14 @@ AC_MSG_CHECKING([for libtiff])
 if test "x${with_libtiff}" = "xyes" -o "x${with_libtiff}" = "x" ; then
 
   dnl Only automatically pick up the external libtiff if it is >= 4.0.
-  AC_CHECK_LIB(tiff,TIFFScanlineSize64,TIFF_SETTING=external HAVE_BIGTIFF=yes,TIFF_SETTING=internal HAVE_BIGTIFF=yes,)
+  PKG_PROG_PKG_CONFIG([0.21])
+  PKG_CHECK_MODULES([TIFF], [libtiff-4 >= 4.0.0], [TIFF_SETTING=external HAVE_BIGTIFF=yes], [TIFF_SETTING=internal HAVE_BIGTIFF=yes])
+  TIFF_SETTING=external
+  LIBS="$TIFF_LIBS $LIBS"
+  CFLAGS="$TIFF_CFLAGS $CFLAGS"
+  EXTRA_INCLUDES="$TIFF_CFLAGS $EXTRA_INCLUDES"
 
-  if test "$TIFF_SETTING" = "external" ; then
-    dnl Cygwin takes a somewhat restrictive view of what should be exported
-    dnl from the dll, so don't use the external library if missing semi-private
-    dnl functions.
-    AC_CHECK_LIB(tiff,_TIFFsetDoubleArray,TIFF_SETTING=external,TIFF_SETTING=internal,)
-  fi
 
-  if test "$TIFF_SETTING" = "external" ; then
-    LIBS="-ltiff $LIBS"
-    AC_MSG_RESULT([using pre-installed libtiff.])
-  else
-    AC_MSG_RESULT([using internal TIFF code.])
-  fi
 
 elif test "x${with_libtiff}" = "xno" ; then
 
@@ -2042,7 +2036,12 @@ if test "$with_geotiff" = "yes" -o "$with_geotiff" = "" ; then
     GEOTIFF_SETTING=internal
   else
     dnl We now require libgeotiff 1.5.0
-    AC_CHECK_LIB(geotiff,GTIFAttachPROJContext,GEOTIFF_SETTING=external,GEOTIFF_SETTING=internal)
+    PKG_PROG_PKG_CONFIG([0.21])
+    PKG_CHECK_MODULES([GEOTIFF], [libgeotiff >= 1.5.0], [GEOTIFF_SETTING=external], [GEOTIFF_SETTING=internal])
+    LIBS="$GEOTIFF_LIBS $LIBS"
+    CFLAGS="$GEOTIFF_CFLAGS $CFLAGS"
+    EXTRA_INCLUDES="$GEOTIFF_CFLAGS $EXTRA_INCLUDES"
+    GEOTIFF_INCLUDE="$GEOTIFF_CFLAGS"
 
   fi
 
@@ -4842,7 +4841,8 @@ dnl ---------------------------------------------------------------------------
 AC_ARG_WITH(libjson-c,[  --with-libjson-c[=ARG]       Include libjson-c support (ARG=internal or libjson-c directory)],,)
 
 if test "$with_libjson_c" = "external" -o "$with_libjson_c" = "" -o "$with_libjson_c" = "yes" ; then
-  AC_CHECK_LIB(json-c,json_object_set_serializer,LIBJSONC_SETTING=external,LIBJSONC_SETTING=internal,)
+  PKG_PROG_PKG_CONFIG([0.21])
+  PKG_CHECK_MODULES([LIBJSONC], [json-c >= 0.1.0], [LIBJSONC_SETTING=external], [LIBJSONC_SETTING=internal])
 elif test "$with_libjson_c" = "internal" ; then
   LIBJSONC_SETTING=internal
 elif test "$with_libjson_c" != "no"; then
@@ -4854,16 +4854,8 @@ fi
 
 if test "$LIBJSONC_SETTING" = "external" ; then
   AC_MSG_RESULT([using pre-installed libjson-c])
-  LIBS="-ljson-c $LIBS"
-  if test "$with_libjson_c" != "" -a "$with_libjson_c" != "yes" -a "$with_libjson_c" != "external" ; then
-    JSON_INCLUDE="-I$with_libjson_c/include/json-c"
-  elif test -f "/usr/include/json-c/json.h"; then
-    JSON_INCLUDE="-I/usr/include/json-c"
-  elif test -f "/usr/local/include/json-c/json.h"; then
-    JSON_INCLUDE="-I/usr/local/include/json-c"
-  else
-    AC_MSG_ERROR([could not find json-c/json.h])
-  fi
+  LIBS="$LIBJSONC_LIBS $LIBS"
+  JSON_INCLUDE="$LIBJSONC_CFLAGS"
 else
     JSON_INCLUDE="-I\$(GDAL_ROOT)/ogr/ogrsf_frmts/geojson/libjson"
     AC_MSG_RESULT([using internal libjson-c code])
